### PR TITLE
Snap 1833

### PR DESCRIPTION
**Pull request by rishitesh**
## What changes were proposed in this pull request?

Added a fallback path for WholeStageCodeGenRDD. As we dynamically change the classloader,  generated code compile time classloaders and runtime class loader might be different. There is no clean way to handle this apart from recompiling the generated code.
This code path will be executed only in case of components having dynamically changing class loaders i.e Snappy jobs & UDFs.  Other sql queries won't be impacted by this.  

## How was this patch tested?

pre-checkin

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 1]
```diff
@@ -916,6 +916,10 @@ object CodeGenerator extends Logging {
     cache.get(code)
   }
 
+  def refresh(code: CodeAndComment) : Unit = {
+    cache.refresh(code)
+  }
+
```
Comment (sumwale) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: Why refresh and not invalidate? The refresh looks a dicey to me (async loading with previous value still valid etc).

File: sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala [hunk 2]
```diff
@@ -500,6 +499,19 @@ case class WholeStageCodegenRDD(@transient sc: SparkContext, var source: CodeAnd
 
   override def compute(split: Partition,
       context: TaskContext): Iterator[InternalRow] = {
+    try {
+      computeInternal(split, context)
+    } catch {
+      case ex: ClassCastException => {
+        // Any other exception in future might be added here
```
Comment (sumwale) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala]: unnecessary brace

File: sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala [hunk 3]
```diff
@@ -500,6 +499,19 @@ case class WholeStageCodegenRDD(@transient sc: SparkContext, var source: CodeAnd
...
+        CodeGenerator.refresh(source)
+        computeInternal(split, context)
```
Comment (sumwale) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala]: What implications does this internal retry have in case the exception comes later down in processing chain? For example if this is a bulk insert/update plan and ClassCastException is something different? Are we expecting the CCE to come at the very start of processing? I guess that should be true for the cases we are attempting to correct, but what about false positives?

**Review by sumwale — COMMENTED** [review]
See some comments. Think about false positives and what implications it can have. If possible reduce false positives, and also discuss with @suranjan about the effect on snapshot TX in progress in the retry.

File: core/src/main/scala/org/apache/spark/executor/Executor.scala [hunk 4]
```diff
@@ -64,7 +64,7 @@ private[spark] class Executor(
 
   private val EMPTY_BYTE_BUFFER = ByteBuffer.wrap(new Array[Byte](0))
 
-  private val conf = env.conf
+  protected val conf = env.conf
```
Comment (sumwale) [inline core/src/main/scala/org/apache/spark/executor/Executor.scala:67]: Why this change?

Comment (rishitesh) [reply core/src/main/scala/org/apache/spark/executor/Executor.scala:67]: We are calling the fetchFile() method based on the jars in the job local variables ( A thread local which is available accross all partitions of a job). fetchFile() needs this value.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala (hunk 2 above)
Comment (rishitesh) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala]: Will remove it.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala (hunk 3 above)
Comment (rishitesh) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala]: If we get a ClasssCastException in any other case, that means it's a bug in the plan and it won't be corrected even if we recompile the plan. And we will get the CCE even after the recompile. This retry will only safeguard those problems which arise due to the different classloader.  
Or, did I misunderstood your question ?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala (hunk 1 above)
Comment (rishitesh) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: OK. Will change the call. Yes, refresh might cause some issues.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala (hunk 3 above)
Comment (rishitesh) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala]: Also, the cases this will solve only the case where a plan is trying to get an object from input adapter row.  e.g. com.blah.Event event = (com.blah.Event) inputRow.get(0).
Even if the Event class has some complex object structure and contains other objects which might be a potential candidate for CCE, that will happen before processing any row. Hence I feel it would be safe to assume not a single row has been processed.

Comment (sumwale) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala]: > And we will get the CCE even after the recompile.

Yes, question was that those cases of genuine CCEs will lead to exactly one retry that can mean duplicates etc. For example a bug in user type impl. Anyway this was meant to ponder whether this should do for all CCEs or identify the specific ClassLoader case by looking at the message also (which should also look like saying cast to same type failed).

Comment (rishitesh) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala]: I am not sure how we can identify same class cases in CCE. Also, commit or rollback will happen in partition level. Even if we retry for a false positive, it will either a complete partition or none. Isn't it?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala [hunk 5]
```diff
@@ -479,8 +482,7 @@ case class CollapseCodegenStages(conf: SQLConf) extends Rule[SparkPlan] {
   }
 }
 
-
-case class WholeStageCodegenRDD(@transient sc: SparkContext, var source: CodeAndComment,
+private[spark] class WholeStageCodegenRDD(@transient sc: SparkContext, var source: CodeAndComment,
```
Comment (hbhanawat) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala]: Any specific reason for changing the it to a non-case class? This class was specifically made a case class to improve concurrency. See https://github.com/SnappyDataInc/spark/pull/63

Comment (rishitesh) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala]: I guess I can revert that, as I was trying to extend this RDD , this was the fallout of that. Now that I am not I can make it as a case class again.
